### PR TITLE
Dynamically infer classpath by classloader according to the class usage

### PR DESCRIPTION
**Pull request by gaojieliu**
When application is using play framework with hot-reload enabled, fast-avro
couldn't find the right classpath by extracting from system property:java.class.path.
So this code change will explicitly find all the classes being used in the
generated files and put the corresponding libraries in the compile classpath.
Play framework is using both current classloader and the thread context classloader
to discover class path in hot reload feature. In fast-avro class generation/compilation,
we have to do the same time to align with this behavior.
Once the generated file is compiled successfully, loading the compiled classes
is not a problem in play framework with hot-reload enabled.

### CODE REVIEW

**Comment by gaojieliu** [comment]
@FelixGV @radai-rosenblatt 
Please review.

File: avro-fastserde/gettingstarted.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+fastSpecificDatumReader.read(reuse, binaryDecoder);
+
+FastSpecificDatumWriter<T> fastSpecificDatumWriter = new FastSpecificDatumWriter<>(schema);
+fastSpecificDatumWriter.write(data, binaryEncoder);
+
+
+## Troubleshooting
+
+
+## Related Documentation
```
Comment (FelixGV) [inline avro-fastserde/gettingstarted.md]: I think we should remove this instead of leaving those sections empty.

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerializerGeneratorBase.java [hunk 2]
```diff
@@ -60,22 +61,25 @@ protected static int nextUniqueInt() {
   public abstract FastSerializer<T> generateSerializer();
 
   @SuppressWarnings("unchecked")
-  protected Class<FastSerializer<T>> compileClass(final String className) throws IOException, ClassNotFoundException {
+  protected Class<FastSerializer<T>> compileClass(final String className, Set<String> knownUsedFullyQualifiedClassNameSet)
```
Comment (FelixGV) [inline avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerializerGeneratorBase.java]: Looks like this class and `FastDeserializerGeneratorBase` have a lot in common. We should consider refactoring the code so they share a common base class or share utility functions...

I don't think it's critical to do this as part of this commit, but we should at least leave a TODO comment.

I leave it up to you whether you want to refactor right away or just leave the TODO (unless you disagree and think the two classes ought to stay separate?)

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java [hunk 3]
```diff
@@ -159,7 +159,7 @@ public Float get(int i) {
    *
    * TODO: verify if it's enough to simply store the data as primitives in order to benefit from the GC optimization.
    */
-  private boolean addPrimitive(float o) {
+  public boolean addPrimitive(float o) {
```
Comment (FelixGV) [inline avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java]: Why did we need this change? It doesn't seem related to this commit, right?

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java [hunk 4]
```diff
@@ -25,20 +25,32 @@
...
+   */
+  private final Set<String> fullyQualifiedClassNameSet = new HashSet<>();
 
   public SchemaAssistant(JCodeModel codeModel, boolean useGenericTypes) {
     this.codeModel = codeModel;
     this.useGenericTypes = useGenericTypes;
     this.exceptionsFromStringable = new HashSet<>();
   }
 
+  protected Set<String> getUsedFullyQualifiedClassNameSet() {
```
Comment (FelixGV) [inline avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java]: Doesn't seem really "schema related". Should we move this functionality to `Utils` instead?

File: avro-fastserde/gettingstarted.md (hunk 1 above)
Comment (gaojieliu) [reply avro-fastserde/gettingstarted.md]: Sure, I can do that.

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerializerGeneratorBase.java (hunk 2 above)
Comment (gaojieliu) [reply avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerializerGeneratorBase.java]: You are right, and this situation has been there from the beginning, and we haven't optimized the code structure yet.
I could leave a TODO.

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java (hunk 3 above)
Comment (gaojieliu) [reply avro-fastserde/src/main/java/com/linkedin/avro/fastserde/PrimitiveFloatList.java]: This is required by LinkedIn internal application.

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java (hunk 4 above)
Comment (gaojieliu) [reply avro-fastserde/src/main/java/com/linkedin/avro/fastserde/SchemaAssistant.java]: This set is used to capture all the used classes during code generation (SchemaAssistant#classFromSchema is the function to extract all the used classes), and I think this is the right place to get this piece of info.
